libs/mrwebrtc/src/media/audio_track_read_buffer.cpp

```diff
@@ -145,49 +138,78 @@ void AudioTrackReadBuffer::Buffer::addFrame(const Frame& frame,
+  float* dst = data_out;
+  const int data_len_orig = *data_len_in_out;
+  int dst_len = data_len_orig;  // number of points remaining
+
+  *has_overrun_out = false;
 
-  while (dstLen > 0) {
-    if (sampleRate == buffer_.rate_ && channels == buffer_.channels_ &&
+  while (dst_len > 0) {
+    if (sample_rate == buffer_.rate_ && num_channels == buffer_.channels_ &&
         buffer_.available()) {
-      // format matches, fill some from the buffer. If the format doesn't
-      // match we will fall through and ensure the next frame matches. This
-      // may drop some data but will only happen when the output
-      // sampleRate/channels change (i.e. rarely)
-      int len = buffer_.readSome(dst, dstLen);
+      // There is still data in the buffer and the format matches, read some.
+      int len = buffer_.readSome(dst, dst_len);
       dst += len;
-      dstLen -= len;
+      dst_len -= len;
     } else {
-      Frame frame;
+      // Buffer is empty or the format has changed.
+      // If the format doesn't match we drop what's left in the buffer and
+      // ensure the next frame matches. This may drop some data but will only
+      // happen when the output sample rate/channels change (i.e. rarely)
+
+      std::optional<Frame> frame;
       {
         std::unique_lock<std::mutex> lock(frames_mutex_);
-        if (frames_.empty()) {  // no more input! fill with sin wave
-          lock.unlock();
-          constexpr float freq = 2 * 222 * float(M_PI);
-          for (int i = 0; i < dstLen; ++i) {
-            dst[i] = 0.15f * sinf((freq * (sinwave_iter_ + i)) /
-                                  (sampleRate * channels));
-          }
-          sinwave_iter_ = (sinwave_iter_ + dstLen) % 628318530 /*twopi*/;
-          sinwave_iter_ += dstLen;
-          return;  // and return
+
+        // Read and reset the overrun flag.
+        *has_overrun_out = *has_overrun_out || has_overrun_;
+        has_overrun_ = false;
+
+        // Pop the next frame.
+        if (!frames_.empty()) {
+          frame = std::move(frames_.front());
+          frames_.pop_front();
         }
-        Frame& f = frames_.front();
-        frame.audio_data.swap(f.audio_data);
-        frame.bits_per_sample = f.bits_per_sample;
-        frame.sample_rate = f.sample_rate;
-        frame.number_of_channels = f.number_of_channels;
-        frame.number_of_frames = f.number_of_frames;
-        frames_.pop_front();
       }
-      buffer_.addFrame(frame, sampleRate, channels);
+
+      if (frame) {
+        buffer_.addFrame(*frame, sample_rate, num_channels);
+      } else {
+        // no more input! fill with sin wave
+        constexpr float freq = 2 * 222 * float(M_PI);
+        switch (pad_behavior) {
+          case mrsAudioTrackReadBufferPadBehavior::kDoNotPad:
+            break;
+          case mrsAudioTrackReadBufferPadBehavior::kPadWithZero:
+            for (int i = 0; i < dst_len; ++i) {
+              dst[i] = 0;
+            }
```

Comment:
`memset(dst, 0, dst_len * sizeof(float))` ?
I know it most likely compile to the same, but be lazy! ;)

Reply:
Fixed